androidApp/src/main/java/com/example/tasky/android/agenda/viewmodel/AgendaViewModel.kt

```diff
@@ -97,16 +101,21 @@ class AgendaViewModel(
     }
 
     private fun getAgendas(timeStamp: Long) {
-        viewModelScope.launch {
-            agendaRepository.getAgenda(timeStamp = timeStamp).onSuccess { data ->
-                val itemList = data.sortedBy { it.getStartTime() }
-                val index = getTimeNeedleDisplayIndex(itemList)
-                val itemUiList: MutableList<AgendaItemUi> = itemList.map { AgendaItemUi.Item(it) }.toMutableList()
-                itemUiList.add(index, AgendaItemUi.Needle)
-
-                _screenStateFlow.update { it.copy(agendas = itemUiList.toImmutableList()) }
+        getAgendasJob?.cancel()
+
+        getAgendasJob =
+            viewModelScope.launch {
+                agendaRepository.getAgendaFlow(timeStamp = timeStamp).onSuccess { flow ->
+                    flow.collectLatest { data ->
+                        val itemList = data.sortedBy { it.getStartTime() }
+                        val index = getTimeNeedleDisplayIndex(itemList)
+                        val itemUiList: MutableList<AgendaItemUi> = itemList.map { AgendaItemUi.Item(it) }.toMutableList()
+                        itemUiList.add(index, AgendaItemUi.Needle)
+
+                        _screenStateFlow.update { it.copy(agendas = itemUiList.toImmutableList()) }
+                    }
+                }
```

Comment:
I would suggest returning a `Flow` directly from your repository. You have the `catch {}` operator if you want to handle any errors. The above can then be changed to:

```
// All in your ViewModel
private val selectedTime = MutableStateFlow(0L)

init {
  selectedTime
    .flatMapLatest {
      agendaRepository.getAgendaFlow(timeStamp = it)
    }
    .onEach { data -> 
      /* Process the list and update your state */
    }
    .launchIn(viewModelScope)
}
```

Besides the simplified `Flow`, I’ve also included a private state called `selectedTime` which, when changed, will trigger the logic in the `init` block which will start (or restart) the flow to feetch the agenda items and trigger the `onEach` block where you will update your state.

`Flow`s also have a great function called `launchIn()` which acts the same as `scope.launch { /* collect flow here */ }`, but without the explicit flow collection and without the extra indentation of the code.